src/examples/TDXContainers.sol

```diff
@@ -0,0 +1,189 @@
+        require(msg.sender == owner);
+
+        require(quoteMeasurementHash == bytes32(0)); // Only once
+
+        bytes1 quote_version = quote[0];
+        bytes memory quoteMeasurement = new bytes(144);
+        bytes1[48] memory rtmr3;
+        if (quote_version == 0x04) {
+            for (uint i = 184; i < 184+48; i++) { /* mrTd */
+                quoteMeasurement[i-184] = quote[i];
+            }
+            for (uint i = 376; i < 376+144; i++) { /* rtmr 0-2 */
+                quoteMeasurement[i+32-376] = quote[i];
+            }
+            for (uint i = 376+144; i < 376+144+48; i++) {
+                rtmr3[i-376-144] = quote[i];
+            }
+        } else if (quote_version == 0x05) {
+            /* TODO: check body type (doesnt matter for now) */
+            for (uint i = 184+6; i < 184+6+48; i++) { /* mrTd */
+                quoteMeasurement[i-184-6] = quote[i];
+            }
+            for (uint i = 376+6; i < 376+6+144; i++) { /* rtmr 0-2 */
+                quoteMeasurement[i+32-376-6] = quote[i];
+            }
+            for (uint i = 376+6+144; i < 376+6+144+48; i++) {
+                rtmr3[i-376-144-6] = quote[i];
+            }
+        } else revert("unsupported quote version");
+
+        bytes32 _quoteMeasurementHash = keccak256(quoteMeasurement);
+        /* TODO: verify measurements are as expected! */
+        /* TODO: require(_quoteMeasurementHash == _expectedMeasurement); */
+
+        bytes32 hostPubkeyHash;
+        assembly {
+            hostPubkeyHash := mload(add(rtmr3, 32))
+        }
+        require(hostPubkeyHash == keccak256(_hostPubkey));
+
+        keymgr.Suave().verifyTDXDCAPQuote(quote, pckCertPem, pckCrlPem, tcbInfo, qeIdentity);
+
+        return  (_quoteMeasurementHash, _hostPubkey, keymgr.attest(keccak256(abi.encodePacked("hPub", _hostPubkey, "qM", _quoteMeasurementHash))));
+    }
+
+    /* Offchain. Decrypts data using keymgr's key and encrypts it to host's pubkey */
+    function reencryptToHostKey(bytes memory ciphertext /* TODO: should we also require a specific quote measurement? */) external returns (bytes memory) {
+        bytes memory plaintext = PKE.decrypt(keymgr.derivedPriv(), ciphertext);
+        return PKE.encrypt(hostPubkey, keymgr.Suave().localRandom(), plaintext);
+    }
+
+    event NewContainerRequested(bytes32, ContainerConfig);
+    function submit(ContainerConfig calldata container) external {
+        bytes32 id = keccak256(abi.encode(container));
+        pendingContainers[id] = container;
+        emit NewContainerRequested(id, container);
+    }
+
+    function approve(bytes32 id) external {
+        require(msg.sender == owner);
+
+        ContainerConfig memory container = pendingContainers[id];
+        if (container.imageHash == bytes32(0)) revert("unknown container");
+
+        approvedContainers[id] = container;
+    }
+
+    event StartContainerRequested(bytes32, ContainerConfig);
+    function start(bytes32 id) external {
```

Comment:
This is cool, having the start, stop, and heartbeats tracked on chain, like a cloud w/ publicly visible dashboard and all the interactions logged